### DISCUSSION TITLE
Exceptions while starting and running the 5.1.3

### DISCUSSION THREAD

**Discussion by ali3n7** [Pryaxis/TShock#2935] [General]
Hi, 
very new here. I was able to install the 5.1.3 on my Raspberry and finally was able to create new worlds. Everything worked fine, unless I created a new middle sized world.
Joining the world was possible, the first save operation threw the following expeption.

`Error on message Terraria.MessageBuffer System.DataMisalignedException: A datatype misalignment was detected in a load or store instruction. at Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?729210>(NPC , Double& , Int32 , Single , Int32 , Boolean , Boolean , Boolean , Entity ) at Terraria.MessageBuffer.GetData(Int32 start, Int32 length, Int32& messageType) at Terraria.NetMessage.CheckBytes(Int32 bufferIndex) Error on message Terraria.MessageBuffer System.DataMisalignedException: A datatype misalignment was detected in a load or store instruction. at Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?729210>(NPC , Double& , Int32 , Single , Int32 , Boolean , Boolean , Boolean , Entity ) at Terraria.MessageBuffer.GetData(Int32 start, Int32 length, Int32& messageType) at Terraria.NetMessage.CheckBytes(Int32 bufferIndex) Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt. at System.Runtime.CompilerServices.CastHelpers.IsInstanceOfClass(Void*, System.Object) at TerrariaApi.Server.Hooking.NpcHooks.OnStrike(orig_StrikeNPC, Terraria.NPC, Int32, Single, Int32, Boolean, Boolean, Boolean, Terraria.Entity) at DynamicClass.Hook<Terraria.NPC::StrikeNPC>?729210(Terraria.NPC, Int32, Single, Int32, Boolean, Boolean, Boolean, Terraria.Entity) at DynamicClass.Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?729210>(Terraria.NPC, Double ByRef, Int32, Single, Int32, Boolean, Boolean, Boolean, Terraria.Entity) at Terraria.MessageBuffer.GetData(Int32, Int32, Int32 ByRef) at Terraria.NetMessage.CheckBytes(Int32) at Terraria.Netplay.UpdateServerInMainThread() at Terraria.Netplay.UpdateInMainThread() at Terraria.Main.DoUpdate(Microsoft.Xna.Framework.GameTime ByRef) at Terraria.Main.Update(Microsoft.Xna.Framework.GameTime) at Terraria.Main.DedServ() at Terraria.Program.orig_RunGame() at Terraria.Program.RunGame() at Terraria.Program.LaunchGame(System.String[], Boolean) at Terraria.WindowsLaunch.orig_Main(System.String[]) at Terraria.WindowsLaunch.Main(System.String[]) at TerrariaApi.Server.Program.StartServer(System.String[]) at TerrariaApi.Server.Program.Main(System.String[]) at Program+<>c__DisplayClass0_0.<<Main>$>g__Start|1() at Program.<Main>$(System.String[]) `

I restarted the Pi and now starting TShock.Server throws also an exception. :(
`Error on message Terraria.MessageBuffer System.DataMisalignedException: A datatype misalignment was detected in a load or store instruction. at Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?729210>(NPC , Double& , Int32 , Single , Int32 , Boolean , Boolean , Boolean , Entity ) at Terraria.MessageBuffer.GetData(Int32 start, Int32 length, Int32& messageType) at Terraria.NetMessage.CheckBytes(Int32 bufferIndex) Error on message Terraria.MessageBuffer System.DataMisalignedException: A datatype misalignment was detected in a load or store instruction. at Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?729210>(NPC , Double& , Int32 , Single , Int32 , Boolean , Boolean , Boolean , Entity ) at Terraria.MessageBuffer.GetData(Int32 start, Int32 length, Int32& messageType) at Terraria.NetMessage.CheckBytes(Int32 bufferIndex) Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt. at System.Runtime.CompilerServices.CastHelpers.IsInstanceOfClass(Void*, System.Object) at TerrariaApi.Server.Hooking.NpcHooks.OnStrike(orig_StrikeNPC, Terraria.NPC, Int32, Single, Int32, Boolean, Boolean, Boolean, Terraria.Entity) at DynamicClass.Hook<Terraria.NPC::StrikeNPC>?729210(Terraria.NPC, Int32, Single, Int32, Boolean, Boolean, Boolean, Terraria.Entity) at DynamicClass.Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?729210>(Terraria.NPC, Double ByRef, Int32, Single, Int32, Boolean, Boolean, Boolean, Terraria.Entity) at Terraria.MessageBuffer.GetData(Int32, Int32, Int32 ByRef) at Terraria.NetMessage.CheckBytes(Int32) at Terraria.Netplay.UpdateServerInMainThread() at Terraria.Netplay.UpdateInMainThread() at Terraria.Main.DoUpdate(Microsoft.Xna.Framework.GameTime ByRef) at Terraria.Main.Update(Microsoft.Xna.Framework.GameTime) at Terraria.Main.DedServ() at Terraria.Program.orig_RunGame() at Terraria.Program.RunGame() at Terraria.Program.LaunchGame(System.String[], Boolean) at Terraria.WindowsLaunch.orig_Main(System.String[]) at Terraria.WindowsLaunch.Main(System.String[]) at TerrariaApi.Server.Program.StartServer(System.String[]) at TerrariaApi.Server.Program.Main(System.String[]) at Program+<>c__DisplayClass0_0.<<Main>$>g__Start|1() at Program.<Main>$(System.String[]) `

Any hints to that? Thanks for any help. Alex

**Comment by sgkoishi** [comment]
Are you running 32-bit? Try run `which dotnet | xargs -d '\n' file`

**Reply by ali3n7** [reply]
yes 32bit

the command gives me the following result: /usr/local/bin/dotnet: symbolic link to /opt/dotnet/dotnet

Next try strated the server again. Once starting playing the again exceptions, but without terminating the server. 
`Error on message Terraria.MessageBuffer
System.NullReferenceException: Object reference not set to an instance of an object.
   at TerrariaApi.Server.Hooking.NpcHooks.OnStrike(orig_StrikeNPC orig, NPC npc, Int32 Damage, Single knockBack, Int32 hitDirection, Boolean crit, Boolean noEffect, Boolean fromNet, Entity entity) in /home/runner/work/TShock/TShock/TerrariaServerAPI/TerrariaServerAPI/TerrariaApi.Server/Hooking/NpcHooks.cs:line 54
`

**Reply by ali3n7** [reply]
here the dotnet info

.NET SDK:
 Version:   6.0.407
 Commit:    d8659ea151

Laufzeitumgebung:
 OS Name:     raspbian
 OS Version:  10
 OS Platform: Linux
 RID:         linux-arm
 Base Path:   /opt/dotnet/sdk/6.0.407/

global.json file:
  Not found

Host:
  Version:      6.0.15
  Architecture: arm
  Commit:       5edef4b20b

.NET SDKs installed:
  6.0.407 [/opt/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.15 [/opt/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.15 [/opt/dotnet/shared/Microsoft.NETCore.App]

**Reply by sgkoishi** [reply]
And yes, 32-bit is not working. It is likely that the upstream had it fixed but didn't make a release, so you might need to compile some of the dependencies (and toolchain) to make it work on 32-bit.
Or, switch to 64-bit.

If you already have a 64-bit OS but still running 32-bit dotnet, download the latest experimental version from the Actions (it comes with an installer to download things for you).

**Reply by ali3n7** [reply]
sorry for bothering you, but can you give me a start how to compile or whatever? I'm really a newbie in that area and I doubt that I get it done. THX

**Reply by sgkoishi** [reply]
It's suggested to switch to a 64-bit OS :eyes:
The compile path requires compiling patchers and patches and running the patchers to patch and compile other things against the patched things - just don't do it.

**Reply by ali3n7** [reply]
good hint :) Update to 64Bit was easy.
updated .NET
downloaded the 64bit ARM version of 5.1.3
using ./TShock.Server gives me the error now: file or directory not found
any hints?

`pi@SYS-Server-Raspi4:~/tshock $ ./TShock.Server
bash: ./TShock.Server: Datei oder Verzeichnis nicht gefunden
`

**Reply by sgkoishi** [reply]
If you know the file is there, this likely means you have a 32-bit dotnet.
> If you already have a 64-bit OS but still running 32-bit dotnet, download the latest experimental version from the Actions (it comes with an installer to download things for you).

**Reply by Arthri** [reply]
Check the output of `file "$(which dotnet)"`. That should tell you what architecture of .NET you have. If it's not ARM64, uninstall it and get the ARM64 version

**Reply by ali3n7** [reply]
I fear I messed something up while "uninstalling" the arm version which I was forced to do manually. Searched for hours and only found the manual cleanup and purge the various directories.
Finally I got the arm64 installed with the install script, but no I get the following
I'm really done and have no clue anymore. Thanks for any help

![image](https://user-images.githubusercontent.com/129767791/229897621-0533dfbe-1109-4ad3-8a6f-78983e4ed2ca.png)

![image](https://user-images.githubusercontent.com/129767791/229897861-c066fb75-5524-4730-81cd-29852a9d7dba.png)

**Reply by sgkoishi** [reply]
_Download the [latest experimental version of TShock from the Actions](https://github.com/Pryaxis/TShock/actions/workflows/ci-otapi3.yml?query=branch%3Ageneral-devel), it comes with an installer `TShock.Installer`._

**Reply by ali3n7** [reply]
I think I still have a problem with executing. Tried ./Thock.Installer and still get the message not found

![image](https://user-images.githubusercontent.com/129767791/229990383-7b97dbab-f3a6-40c3-80eb-d9a1d1d5e3a9.png)

**Reply by sgkoishi** [reply]
Can you run `uname -a` and `file ./TShock.Installer`?

**Reply by ali3n7** [reply]
`pi@SYS-Server-Raspi4:~/tshock $ uname -a
Linux SYS-Server-Raspi4 5.10.103-v8+ #1529 SMP PREEMPT Tue Mar 8 12:26:46 GMT 2022 aarch64 GNU/Linux
pi@SYS-Server-Raspi4:~/tshock $ file ./TShock.Installer
./TShock.Installer: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=72b7d6d702a72abbc534c407850a5596656ac584, stripped`

![image](https://user-images.githubusercontent.com/129767791/230029388-d8e6274f-b824-4d8b-9a75-2d63ae788e91.png)

**Reply by sgkoishi** [reply]
Hmmmmm... perhaps some random dependencies missing? try `ldd TShock.Installer`?

**Reply by ali3n7** [reply]
program not dynamically linked ...hmmmm

**Comment by ali3n7** [comment]
after a lot of attempts and crawling through the net, I just recognized, that I only switched the kernel mode to 64bit, but not the OS. So the 64 bit dotnet will never work on it. Was able to revert all changes and start from scratch.
I use now the latest arm build from Actions, but still get the following error. 
Can I commit that error? If so, how? Thanks.
Nonetheless, I will do an fully new and additional installation with 64bit OS on a new SD card and try to run the server from there.

Error on message Terraria.MessageBuffer
System.NullReferenceException: Object reference not set to an instance of an object.
   at TerrariaApi.Server.Hooking.NpcHooks.OnStrike(orig_StrikeNPC orig, NPC npc, Int32 Damage, Single knockBack, Int32 hitDirection, Boolean crit, Boolean noEffect, Boolean fromNet, Entity entity) in /home/runner/work/TShock/TShock/TerrariaServerAPI/TerrariaServerAPI/TerrariaApi.Server/Hooking/NpcHooks.cs:line 54
   at Hook<Terraria.NPC::StrikeNPC>?37629530(NPC , Int32 , Single , Int32 , Boolean , Boolean , Boolean , Entity )
   at Glue:ThiscallStructRetPtr<Terraria.NPC::StrikeNPC,Hook<Terraria.NPC::StrikeNPC>?37629530>(NPC , Double& , Int32 , Single , Int32 , Boolean , Boolean , Boolean , Entity )
   at Terraria.MessageBuffer.GetData(Int32 start, Int32 length, Int32& messageType)
   at Terraria.NetMessage.CheckBytes(Int32 bufferIndex)
